src/core/control/tools/Selection.cpp

```diff
@@ -153,19 +178,40 @@ auto RegionSelect::contains(double x, double y) const -> bool {
     return (hits & 1) != 0;
 }
 
-auto RegionSelect::finalize(PageRef page) -> bool {
+auto RegionSelect::finalize(PageRef page) -> size_t {
     this->page = page;
-
-    Layer* l = page->getSelectedLayer();
-    for (Element* e: l->getElements()) {
-        if (e->isInSelection(this)) {
-            this->selectedElements.push_back(e);
+    size_t layerId = 0;
+
+    if (multiLayer) {
+        for (int layerNo = page->getLayers()->size() - 1; layerNo >= 0; layerNo--) {
+            Layer* l = page->getLayers()->at(layerNo);
+            bool selectionOnLayer = false;
+            if (l->isVisible()) {
+                for (Element* e: l->getElements()) {
+                    if (e->isInSelection(this)) {
+                        this->selectedElements.push_back(e);
+                        selectionOnLayer = true;
+                    }
+                }
+            }
+            if (selectionOnLayer) {
+                layerId = layerNo + 1;
+                break;
+            }
+        }
+    } else {
+        Layer* l = page->getSelectedLayer();
+        for (Element* e: l->getElements()) {
+            if (e->isInSelection(this)) {
+                this->selectedElements.push_back(e);
+                layerId = page->getSelectedLayerId();
+            }
         }
     }
 
     this->viewPool->dispatchAndClear(xoj::view::SelectionView::DELETE_VIEWS_REQUEST, this->bbox);
 
-    return !this->selectedElements.empty();
+    return layerId;
 }
```

Comment:
This whole method is a duplication of `RectSelection::finalize`. I suggest to make from these duplicated methods `Selection::finalize` and thus get rid of the duplication.

Reply:
Oh, yes. This is something I wanted to ask you about, but forgot to write into the PR-description. I also think this duplicate code should be merged into one method. I'll do that.